modules/calc/src/main/java/com/opengamma/strata/calc/runner/CalculationParameters.java

```diff
@@ -0,0 +1,244 @@
+/**
+ * Copyright (C) 2016 - present by OpenGamma Inc. and the OpenGamma group of companies
+ *
+ * Please see distribution for license.
+ */
+package com.opengamma.strata.calc.runner;
+
+import static com.opengamma.strata.collect.Guavate.toImmutableMap;
+
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Stream;
+
+import org.joda.beans.BeanDefinition;
+import org.joda.beans.ImmutableBean;
+import org.joda.beans.JodaBeanUtils;
+import org.joda.beans.MetaBean;
+import org.joda.beans.Property;
+import org.joda.beans.PropertyDefinition;
+import org.joda.beans.impl.light.LightMetaBean;
+
+import com.google.common.collect.ImmutableMap;
+import com.opengamma.strata.basics.CalculationTarget;
+import com.opengamma.strata.calc.config.Measure;
+import com.opengamma.strata.collect.MapStream;
+
+/**
+ * The calculation parameters.
+ * <p>
+ * This provides a set of parameters that will be used in a calculation.
+ * Each parameter defines a {@linkplain CalculationParameter#queryType() query type},
+ * thus the functions are keyed in a {@code Map} by the query type {@code Class}.
+ */
+@BeanDefinition(style = "light")
+public final class CalculationParameters implements ImmutableBean, Serializable {
+
+  /**
+   * An empty instance.
+   */
+  private static final CalculationParameters EMPTY = new CalculationParameters(ImmutableMap.of());
+
+  /**
+   * The parameters, keyed by query type.
+   */
+  @PropertyDefinition(validate = "notNull")
+  private final ImmutableMap<Class<?>, CalculationParameter> parameters;
+
+  //-------------------------------------------------------------------------
+  /**
+   * Obtains an empty instance with no parameters.
+   * 
+   * @return the empty instance
+   */
+  public static CalculationParameters empty() {
+    return EMPTY;
+  }
+
+  /**
+   * Obtains an instance from the specified parameters.
+   * <p>
+   * The list will be converted to a {@code Map} using {@link CalculationParameter#queryType()}.
+   * Each parameter must refer to a different query type.
+   * 
+   * @param parameters  the parameters
+   * @return the calculation parameters
+   */
+  public static CalculationParameters of(CalculationParameter... parameters) {
+    if (parameters.length == 0) {
+      return EMPTY;
+    }
+    return new CalculationParameters(Stream.of(parameters).collect(toImmutableMap(p -> p.queryType())));
```

Comment:
Using `MapStream` is a bit neater:

    return new CalculationParameters(MapStream.of(parameters, p -> p.queryType()).toMap();